src/dump.complex.js

```diff
@@ -0,0 +1,100 @@
+'use strict'
+
+const dumpBool = require('./dump.bool')
+const dumpFloat = require('./dump.float')
+const dumpInt = require('./dump.int')
+const dumpNull = require('./dump.null')
+const dumpString = require('./dump.string')
+const dumpSymbol = require('./dump.symbol')
+const dumpSymbolLink = require('./dump.symbol.link')
+
+/**
+ * @param {string} symbol
+ * @param {Array<string>} symcache
+ * @returns {Buffer}
+ */
+const dumpSymbolOrLink = (symbol, symcache) => {
+  const pos = symcache.indexOf(symbol)
+  if (pos === -1) {
+    symcache.push(symbol)
+    return dumpSymbol(symbol)
+  }
+
+  return dumpSymbolLink(pos)
+}
+
+/**
+ * @param {Array} arr
+ * @param {Array<string>} symcache
+ * @returns {Buffer}
+ */
+const dumpArray = (arr, symcache) => {
+  const head = Buffer.from('[', 'utf8')
+  const len = dumpInt(arr.length).slice(1)
+  const data = []
+
+  for (const val of arr) {
+    data.push(encodeType(val, symcache))
+  }
+
+  return Buffer.concat([
+    head,
+    len,
+    ...data
+  ])
+}
+
+/**
+ * @param {Object} obj
+ * @param {Array<string>} symcache
+ * @returns {Buffer}
+ */
+const dumpHash = (obj, symcache) => {
+  const head = Buffer.from('{', 'utf8')
+  const len = dumpInt(Object.keys(obj).length).slice(1)
+
+  const data = []
+  for (const key in obj) {
+    data.push(dumpSymbolOrLink(key, symcache), encodeType(obj[key], symcache))
+  }
+
+  return Buffer.concat([
+    head,
+    len,
+    ...data
+  ])
+}
+
+/**
+ * @param {any} val
+ * @param {Array<string>} symcache
+ * @returns {Buffer}
+ */
+const encodeType = (val, symcache) => { // see http://jakegoulding.com/blog/2013/01/15/a-little-dip-into-rubys-marshal-format
+  let symbol
+
+  switch (typeof val) {
+    case 'string': return dumpString(val)
+    case 'boolean': return dumpBool(val)
+    case 'number':
+      if (Number.isInteger(val)) return dumpInt(val)
+      return dumpFloat(val)
+    case 'object':
```

Comment:
typeof new Map()
"object"
typeof new Set()
"object"

Reply:
typeof new Uint8Array()
"object"

Reply:
throw for all cases where its noll null, a simple array or an object literal?

Reply:
hmm we can add casting to primitive objects in these cases I think, or throw an error in case if it's not plain object?

Reply:
for case of uint I can see that it converts it to a hash on ruby but other two classes are empty objects:
```
2.6.6 :001 > Marshal.load(['04087b083a063069063a063169073a06326908'].pack('H*'))
 => {:"0"=>1, :"1"=>2, :"2"=>3}
```

Reply:
for the first version, i would throw on special object types